### PR TITLE
Monotonic rewrite

### LINKED ISSUE #859: Higher resolution monotonic timer (on stm32)
It seems the resolution of timer based monotonics on the stm32 is a hard limit set by this constant:
`const TIMER_HZ: u32 = 1_000_000;`

Is there a reason for this, or could TIMER_HZ be made a generic? I would like to have a bit more resolution for my use case.

### LINKED ISSUE #873: Monotonics: non-hardcoded tick rate
By @Finomnis in [Matrix chat](https://matrix.to/#/!yafYEipFNsXDdwiHMT:matrix.org/$OHJAn0Qp8YIDLFhhW86z_Y8GbsyIdtcAY-Pcxyo5b3o?via=matrix.org&via=mozilla.org&via=nitro.chat)

--------

I wonder if it would be worth redesigning monotonics to not have a hardcoded tick rate. Then the given input clock frequency could simply become the denominator of the Instant/Duration.

There is also another reason I am interested in a rewrite: the `Monotonic` trait is a weird mix of an internal and an external API. It is intended to be used as an interface for the Timer Queue, which is kind of an internal implementation detail. And yet its `now` function is kind of a public end-user API.

Steps that I think about:

  * make `Monotonic` purely internal
  * make `Monotonic` purely tick passed, no fugit involved
  * create a wrapper struct in the user's code via a macro that then converts the "now" from the tick based monotonic to a fugit based timestamp

We need to proxy the delay functions of the timer queue anyway, so we could simply perform the conversion in those proxy functions.

One more thing- I would like to rewrite all those `if now >= wakeup` to `if signed(now - wakeup) >= 0` to allow for a clean handling of wrapping monotonics.

I know all of this would definitely require a new major version bump of rtic-monotonics and rtic-time.

What are your thoughts?

### PR DESCRIPTION
**Pull request by Finomnis**
See #873.

### CODE REVIEW

**Comment by korken89** [comment]
Hi @Finomnis , sorry for the delay on this!

Overall it looks like a very nice improvement of the monotonics. Would you like to cleanup the PR to get tests to start passing?
And please give us a ping when we can test it more :D

**Comment by Finomnis** [comment]
@korken89 Hey! Sorry for the delay, I was kinda busy with family and stuff. I'll start working on this now again, looking forward to your feedback.

**Comment by Finomnis** [comment]
TODO: Replace all `mod _interrupt` with a nested function in the generated `start` function. That way it's completely hidden.

File: rtic-monotonics/src/imxrt.rs
Comment (Finomnis) [inline rtic-monotonics/src/imxrt.rs]: This mod doesn't have to be exposed; instead, define the function inside of the timer struct. Same for all other monotonics.

**Comment by Finomnis** [comment]
@korken89 Bump :)

**Comment by misterinterrupt** [comment]
Is this planned for merging or is there still discussion or more to do?

**Comment by Finomnis** [comment]
> Is this planned for merging or is there still discussion or more to do?

Planned for merging, waiting for a review.

**Comment by misterinterrupt** [comment]
Awesome, I like what you've done and looking forward to it 

> > Is this planned for merging or is there still discussion or more to do?
> 
> Planned for merging, waiting for a review.

File: rtic-time/src/monotonic.rs
Comment (korken89) [inline rtic-time/src/monotonic.rs]: After playing a bit it feels like we should add an `interval` method here that does the heavy lifting of `timeout_at` and manages the instances. Seems a too common usecase to leave out. :) 

Thoughts?

Specifically I am talking about the following functionality:

https://docs.rs/tokio/latest/tokio/time/fn.interval_at.html
https://docs.rs/tokio/latest/tokio/time/fn.interval.html

Comment (Finomnis) [reply rtic-time/src/monotonic.rs]: I agree :) although this would only be considered a minor change according to semver, so I might suggest we add it afterwards. This pr is already huge as-is. Maybe open an issue for it? Sounds like a good beginner task.

Comment (Finomnis) [reply rtic-time/src/monotonic.rs]: What do you mean with "manages the instances"?

Comment (Finomnis) [reply rtic-time/src/monotonic.rs]: I tried to keep this pr a pure API rework; your suggestion would introduce new functionality.

Comment (korken89) [reply rtic-time/src/monotonic.rs]: Yeah it's fine to not add it :) The extension of functionality we can add later.

**Comment by korken89** [comment]
We just merged a large RISC-V support PR that has minor clashes with this PR, could you rebase?

File: rtic-monotonics/src/nrf/rtc.rs
Comment (korken89) [inline rtic-monotonics/src/nrf/rtc.rs:38]: Making these `pub` is dangerous if we need to bump the PACs. It should be in an `#[doc(hidden)]` module only used to expose them.

File: rtic-monotonics/src/nrf/timer.rs
Comment (korken89) [inline rtic-monotonics/src/nrf/timer.rs:44]: Same comment as previous `pub`

**Review by korken89 — COMMENTED** [review]
Some minor comments

**Comment by Finomnis** [comment]
> Some minor comments

Busy at work RN, might take a look later :)

**Comment by Finomnis** [comment]
@korken89 Fixed

File: rtic-time/Cargo.toml
Comment (AfoHT) [inline rtic-time/Cargo.toml]: Now seems a good time to fix this TODO :)

**Review by AfoHT — COMMENTED** [review]
This is really impressive :)

There's a lot to like here, but in particular how clean the actual usage of monotonics became, awesome!

I know @korken89 got much more experience with monotonics/timer queues so he'll get the last word, but I'm happy to see this PR landing :)

File: rtic-time/Cargo.toml
Comment (Finomnis) [reply rtic-time/Cargo.toml]: Lol :D
Not entirely sure what to write :D

**Comment by AfoHT** [comment]
@Finomnis Could you please rebase on latest master branch and while doing that remove all merge commits?

Might also be a good opportunity to tidy a bit in the commit log 😇

If unsure how, I'm happy to help 👍

**Comment by Finomnis** [comment]
> @Finomnis Could you please rebase on latest master branch and while doing that remove all merge commits?
> 
> Might also be a good opportunity to tidy a bit in the commit log 😇
> 
> If unsure how, I'm happy to help 👍

I planned to do a squash merge, if that's allright. I'm too lazy to clean them up :D

**Comment by burrbull** [comment]
Why have you downgrade esp32c3 version?

**Comment by Finomnis** [comment]
> Why have you downgrade esp32c3 version?

My bad. Merging gets out of hand with these kind of huge merge requests. I hope it is merged soon.

File: rtic-time/Cargo.toml
Comment (Finomnis) [reply rtic-time/Cargo.toml]: Done

**Comment by Finomnis** [comment]
@AfoHT Is it alright with you if we squash merge, instead of me cleaning up the commit history?

**Comment by Finomnis** [comment]
@korken89 Did some minor fixes after merging the main branch, should be stable again

**Comment by AfoHT** [comment]
@Finomnis don't worry about the esp32c3 QEMU run :) 

I'm in the process of squashing/tidying the PR which is about to be merged

But if you got more commits coming I can hold off a little, or we just create another PR. What do you think?

I'll bring in the https://github.com/rtic-rs/rtic/pull/874/commits/986f9594709c936605c7a42b69182d3b3c21b199 onto the squashed/rebased WIP I got :)

**Comment by AfoHT** [comment]
I took the liberty of tacking on a tiny CI fix, to get CI happy enough for a merge :)

**Review by AfoHT — APPROVED** [review]
Awesome work! :tada: Thank you!

I hope I didn't miss anything while squashing, but if that is the case that's on me and we can follow up :+1:

**Comment by SpinFast** [comment]
Very nice

**Pull request merged**
